app/web_conn.go

```diff
@@ -121,45 +131,95 @@ func (wc *WebConn) SetSession(v *model.Session) {
 }
 
 // Pump starts the WebConn instance. After this, the websocket
-// is ready to send/receive messages.
+// is ready to send messages.
 func (wc *WebConn) Pump() {
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		wc.writePump()
-	}()
-	wc.readPump()
-	close(wc.endWritePump)
-	wg.Wait()
+	// writePump is blocking in nature.
+	wc.writePump()
+	// Once it exits, we close everything.
 	wc.App.HubUnregister(wc)
 	close(wc.pumpFinished)
 }
 
-func (wc *WebConn) readPump() {
-	defer func() {
-		wc.WebSocket.Close()
-	}()
-	wc.WebSocket.SetReadLimit(model.SOCKET_MAX_MESSAGE_SIZE_KB)
+// startPoller adds the file descriptor of the connection
+// to the global epoll instance and registers a callback.
+func (wc *WebConn) startPoller() {
+	desc := netpoll.Must(netpoll.HandleRead(wc.WebSocket))
+	wc.App.Srv().Poller().Start(desc, func(wsEv netpoll.Event) {
+		if wsEv&(netpoll.EventReadHup|netpoll.EventHup) != 0 {
+			wc.App.Srv().Poller().Stop(desc)
+			wc.Close()
+			return
+		}
+
+		// Block until we have a token.
+		wc.App.Srv().GetWebConnToken()
```

Comment:
What happens if this returns when the connection/poller has been closed/stopped already?

Reply:
The poller is only stopped in 2 places: L149 and L162 in this file. So if the connection is closed, `ReadMsg()` will fail, and then we stop poller and close connection, and release the token on our way out.

If it happens earlier, then the token is not taken in the first place.